packages/fxa-settings/src/lib/oauth/hooks.tsx

```diff
@@ -0,0 +1,229 @@
+    scope: await integration.getNormalizedScope(),
+  };
+  if (keysJwe) {
+    opts.keys_jwe = keysJwe;
+  }
+  if (integration.data.accessType === 'offline') {
+    opts.access_type = integration.data.accessType;
+  }
+
+  const result = await authClient.createOAuthCode(
+    sessionToken,
+    integration.data.clientId,
+    integration.data.state,
+    opts
+  );
+
+  if (!result) {
+    throw new OAuthErrorInvalidOAuthCodeResult();
+  }
+
+  return result;
+}
+
+/**
+ * Builds an updated redirect url for the relying party
+ * @param oauthCode
+ * @returns
+ */
+function constructOAuthRedirectUrl(
+  oauthCode: {
+    code: string;
+    state: string;
+  },
+  redirectTo: string
+) {
+  // Update the state of the redirect URI
+  const redirectUri = new URL(redirectTo);
+  if (oauthCode.code) {
+    redirectUri.searchParams.set('code', oauthCode.code);
+  }
+  if (oauthCode.state) {
+    redirectUri.searchParams.set('state', oauthCode.state);
+  }
+  return redirectUri;
+}
+
+export type FinishOAuthFlowHandler = (
+  accountUid: string,
+  sessionToken: string,
+  keyFetchToken: string,
+  unwrapKB: string
+) => Promise<{ redirect: string }>;
+
+type FinishOAuthFlowHandlerResult = {
+  finishOAuthFlowHandler: FinishOAuthFlowHandler;
+  oAuthDataError: null | Error;
+};
+
+/**
+ * Generates a redirect link which relays the new oauth token to the relying party.
+ * @param accountUid - Current account uid
+ * @param sessionToken - Current session token
+ * @param keyFetchToken - Current key fetch token
+ * @param unwrapKB - Used to unwrap the account keys
+ * @returns An object containing the redirect URL, that can relay the new OAuthCode.
+ */
+export function useFinishOAuthFlowHandler(
+  authClient: AuthClient,
+  integration: Integration
+): FinishOAuthFlowHandlerResult {
+  const finishOAuthFlowHandler: FinishOAuthFlowHandler = useCallback(
+    async (accountUid, sessionToken, keyFetchToken, unwrapKB) => {
+      const oAuthIntegration = integration as OAuthIntegration;
+      const { kB } = await authClient.accountKeys(keyFetchToken, unwrapKB);
+      const keys = await constructKeysJwe(
+        authClient,
+        oAuthIntegration,
+        accountUid,
+        sessionToken,
+        keyFetchToken,
+        kB
+      );
+      const code = await constructOAuthCode(
+        authClient,
+        oAuthIntegration,
+        sessionToken,
+        keys
+      );
+      const redirectUrl = constructOAuthRedirectUrl(
+        code,
+        oAuthIntegration.data.redirectTo
+      );
+      return {
+        redirect: redirectUrl.href,
+      };
+    },
+    [authClient, integration]
+  );
+
+  // We can't return early because `useCallback` can't be set conditionally
+  if (isOAuthIntegration(integration)) {
+    const oAuthDataError = checkOAuthData(integration);
+    return { oAuthDataError, finishOAuthFlowHandler };
+  }
+  return { oAuthDataError: null, finishOAuthFlowHandler };
```

Comment:
I specified it here because I figured we might be sorting out several different errors in the container components, but yeah, you're probably right. I'll add a note to FXA-8106 👍